community/values/src/test/java/org/neo4j/values/storable/TextValueTest.java

```diff
@@ -0,0 +1,141 @@
+        assertThat( value.apply( "\uD83D\uDE21\uD83D\uDE21\uD83D\uDE21" ).substring( 1, 1 ),
+                equalTo( value.apply( "\uD83D\uDE21" ) ) );
+        assertThat( value.apply( "\uD83D\uDE21\uD83D\uDE21\uD83D\uDE21" ).substring( 1, 2 ),
+                equalTo( value.apply( "\uD83D\uDE21\uD83D\uDE21" ) ) );
+
+        exception.expect( IndexOutOfBoundsException.class );
+        value.apply( "hello" ).substring( -4, 2 );
+    }
+
+    @Test
+    public void toLower()
+    {
+        assertThat( value.apply( "HELLO" ).toLower(), equalTo( value.apply( "hello" ) ) );
+        assertThat( value.apply( "Hello" ).toLower(), equalTo( value.apply( "hello" ) ) );
+        assertThat( value.apply( "hello" ).toLower(), equalTo( value.apply( "hello" ) ) );
+    }
+
+    @Test
+    public void toUpper()
+    {
+        assertThat( value.apply( "HELLO" ).toUpper(), equalTo( value.apply( "HELLO" ) ) );
+        assertThat( value.apply( "Hello" ).toUpper(), equalTo( value.apply( "HELLO" ) ) );
+        assertThat( value.apply( "hello" ).toUpper(), equalTo( value.apply( "HELLO" ) ) );
+    }
+
+    @Test
+    public void ltrim()
+    {
+        assertThat( value.apply( "  HELLO" ).ltrim(), equalTo( value.apply( "HELLO" ) ) );
+        assertThat( value.apply( " Hello" ).ltrim(), equalTo( value.apply( "Hello" ) ) );
+        assertThat( value.apply( "  hello  " ).ltrim(), equalTo( value.apply( "hello  " ) ) );
+        assertThat( value.apply( "\u2009㺂࿝鋦毠\u2009" ).ltrim(), equalTo( value.apply( "㺂࿝鋦毠\u2009" ) ) );
+    }
+
+    @Test
+    public void rtrim()
+    {
+        assertThat( value.apply( "HELLO  " ).rtrim(), equalTo( value.apply( "HELLO" ) ) );
+        assertThat( value.apply( "Hello  " ).rtrim(), equalTo( value.apply( "Hello" ) ) );
+        assertThat( value.apply( "  hello  " ).rtrim(), equalTo( value.apply( "  hello" ) ) );
+        assertThat( value.apply( "\u2009㺂࿝鋦毠\u2009" ).rtrim(), equalTo( value.apply( "\u2009㺂࿝鋦毠" ) ) );
+    }
+
+    @Test
+    public void trim()
+    {
+        assertThat( value.apply( "  hello  " ).trim(), equalTo( value.apply( "hello" ) ) );
+        assertThat( value.apply( "  hello " ).trim(), equalTo( value.apply( "hello" ) ) );
+        assertThat( value.apply( "hello " ).trim(), equalTo( value.apply( "hello" ) ) );
+        assertThat( value.apply( "  hello" ).trim(), equalTo( value.apply( "hello" ) ) );
+        assertThat( value.apply( "\u2009㺂࿝鋦毠\u2009" ).trim(), equalTo( value.apply( "㺂࿝鋦毠" ) ) );
+    }
+
+    @Test
+    public void reverse()
+    {
+        assertThat( value.apply( "Foo" ).reverse(), equalTo( value.apply( "ooF" ) ) );
+        assertThat( value.apply( "" ).reverse(), equalTo( StringValue.EMTPY ) );
+        assertThat( value.apply( " L" ).reverse(), equalTo( value.apply( "L " ) ) );
+        assertThat( value.apply( "\r\n" ).reverse(), equalTo( value.apply( "\n\r" ) ) );
+        assertThat( value.apply( "\uD801\uDC37" ).reverse(), equalTo( value.apply( "\uD801\uDC37" ) ) );
```

Comment:
We cannot see it actually reversing for UTF here. It is valuable to test the single char, but maybe also try a real word?

Reply:
What do you mean by _reversing for UTF_? The last test is testing that we keep the code point `\uD801\uDC37` intact.

Reply:
Aye, but if you have two code points, they should be in reverse order?

Reply:
No, http://unicode.scarfboy.com/?s=U%2B10437

Reply:
Ok, I'm not communicating well.

I understand that the current test check that the single code point `\uD801\uDC37` is not reversed to `\uDC37\uD801`, which would be a completely different unicode thing.

What I'm trying to suggest is to also add a multi-codepoint unicode string, to make sure that we actually can reverse these. E.g. `\uDC37\uD801\uDC2D\uD801 => \uDC2D\uD801\uDC37\uD801` or something.

Reply:
Something like this?